### ISSUE TITLE
Private repository download url gets 404

### ISSUE THREAD

**Issue by jakubmikita** [miya0001/gh-auto-updater#7]
Hi,

thanks for this great lib.

It works perfectly with public repos, but when I try to download an update from private repo I'm getting 404 error. This is because it needs access token passed with a request, but it must be passed as a header if I can see correctly.

Do you maybe know a way how it could be resolved?

**Comment by miya0001** [comment]
@Kubitomakita 

If you check following, you can update the plugin from private repository.

![](https://www.evernote.com/l/ABUZ13BuHKtNALpL763BxBN-lxHMwYApXMYB/image.png)

But It allows full access to your repositories, so you have to create machine user who can't commit to the repositories.

https://developer.github.com/guides/managing-deploy-keys/#machine-users

Actually, I haven't tried this solution it. If it works as expected, please tell me.

Thanks!

**Comment by jakubmikita** [comment]
Hi @miya0001, thanks for your reply.

Without this option checked it wasn't event pulling the repo details from API, so I have it enabled since the beginning.

Please let me rephrase the problem. After repository data is pulled you are getting assets object and the browser_download_url property. This url works only if you have set header with the access token. For public (WordPress updater class) it's returning 404 error.

I found that you can actually query GitHub API asset endpoint and when you have set header "accept: application/octet-stream you" you will get some public zip url. Unfortunately WP Updater doesn't understand it either.

**Comment by miya0001** [comment]
Oh, I understood the situation. It needs `Authentication: xxx` header, right?
I will search for a hook to add a request header when downloading the plugin.

**Comment by jakubmikita** [comment]
Exactly!

I managed to get the asset's AWS url which directly points to the zip but it's encoded by WordPress and it's returning 400... URL is passed between so many classes and functions that I cannot find what the encoding is.

Thanks for your help Takayuki!

**Comment by miya0001** [comment]
I could download an asset from the private repository.
If github token is defined, download url has to be changed.

```
curl -vLJO -H 'Accept: application/octet-stream' \
'https://api.github.com/repos/:owner/:repo/releases/assets/:id?access_token=:token'
```

And there is a filter hook to add `Accept: application/octet-stream`.
https://github.com/WordPress/WordPress/blob/35a0240fb2df290e86157d1d44aa1ccdfb92e228/wp-includes/class-http.php#L223

I'll try to solve this issue tomorrow.
If you have a moment please help me. 😄

**Comment by miya0001** [comment]
Ah, we can download asset from the same url, if token is not defined and the repository is public.
I have to change the download url to the url.

**Comment by jakubmikita** [comment]
I managed to resolve the issue, but it's not-so-clean patch, because I forked it and I have no more time to make it flexible ;)

Please check my code: http://pastebin.com/WHnZ3FWq

**Comment by miya0001** [comment]
Hi @Kubitomakita 

Could you give me a PR for your code.
Basically, It looks good to me.
I want to add some unit tests.

Thanks! 😄

**Comment by AndonMitev** [comment]
Hey, i'm experiencing similar issue 

I do have my url: `https://github.com/AndonMitev/tauri-updater-private-repo/releases/latest/download/latest.json`

it's correct if i switch repo visibility from private to public i'am able to download it

and then i do have the following headers:

```
Authorization: Bearer {PAT},
Accept: application/octet-stream / application/vnd.api+json ( tried both )
``` 
but still hittin 404 and not able to download the json file, any ideas?

PS: Tried with: `https://github.com/AndonMitev/tauri-updater-private-repo/releases/assets/latest.json` again 404
